timeline-chart/src/layer/time-graph-markers-chart.ts

```diff
@@ -0,0 +1,122 @@
+import { TimeGraphMarkerState } from "../components/time-graph-marker";
+import { TimeGraphRow } from "../components/time-graph-row";
+import { TimelineChart } from "../time-graph-model";
+import { TimeGraphRowController } from "../time-graph-row-controller";
+import { TimeGraphChart, TimeGraphChartProviders } from "./time-graph-chart";
+
+/**
+ * Extension of the time-graph-chart, modified to hold logic for the markers layer.
+ * @param id string
+ * @param providers TimeGraphChartProviders
+ * @param rowController TimeGraphRowController
+ */
+export class TimeGraphMarkersChart extends TimeGraphChart {
+
+    private _selectedMarker?: TimelineChart.TimeGraphMarkerState;
+
+    constructor(id: string, protected providers: TimeGraphChartProviders, protected rowController: TimeGraphRowController) {
+        super(id, providers, rowController);
+        
+        this.registerMouseInteractions({
+            click: el => {
+                if (el instanceof TimeGraphMarkerState) {
+                    if (el.model.range.start !== undefined && el.model.range.end !== undefined) {
+                        this.selectedMarker = el.model;
+                        this.unitController.selectionRange = {
+                            start: el.model.range.start,
+                            end: el.model.range.end
+                        };
+                    }
+                }
+            }
+        });
+    }
+
+    protected createNewState = (markerModel: TimelineChart.TimeGraphMarkerState, rowComponent: TimeGraphRow): TimeGraphMarkerState | undefined => {
+        const { xStart, xEnd, displayWidth, elementStyle } = this.getNewStateParams(markerModel);
+        return new TimeGraphMarkerState(markerModel.id, markerModel, xStart, xEnd, rowComponent, elementStyle, displayWidth);
+    }
+
+    onKeydown = (event: KeyboardEvent) => {
```

Comment:
navigation of states are done in time-graph-chart-cursors.ts and done differently. Based on the selected row (y-axis) and time (x-axis) it finds the next state. For markers and in time-graph-chart-cursors.ts, you could loop through all markers for the row and time range and move the cursor to the time of the next marker available. 

An advantage is that you don't need to store next & prev markers in each marker after fetching the marker data which is unnecessary processing. On demand searching for the next marker by looping through all markers of a row should be acceptable.

Reply:
I have implemented this change.  It searches by stateId instead of time, which I think is simpler.